include/elements/2d/quadrilateral_lme_element.tcc

```diff
@@ -78,6 +78,18 @@ inline Eigen::VectorXd mpm::QuadrilateralLMEElement<Tdim>::shapefn(
         }
         J.noalias() += -r * r.transpose();
 
+        //! Add preconditioner for J (Mathieu Foca, PhD Thesis)
+        for (unsigned i = 0; i < Tdim; i++) J.diagonal()[i] += r.norm();
+
+        //! Check reciprocal condition number
+        Eigen::JacobiSVD<Eigen::MatrixXd> svd(J);
+        const double rcond =
+            svd.singularValues()(svd.singularValues().size() - 1) /
+            svd.singularValues()(0);
+        if (rcond < 1E-8)
+          throw std::runtime_error(
+              "Error in QuadLME: The Hessian matrix is singular");
+
```

Comment:
This is a very time costly process to check everything at every element, whereas most of the time it is not needed. Do we really need this? I am not sure what is the benefits of stopping the simulation here.